src/axom/quest/ScatteredInterpolation.hpp

```diff
@@ -244,6 +247,107 @@ class ScatteredInterpolation
   using PointType = typename DelaunayTriangulation::PointType;
   using BoundingBoxType = typename DelaunayTriangulation::BoundingBox;
 
+private:
+  using MortonIndexType = axom::uint64;
+
+  using VertexSet = typename DelaunayTriangulation::IAMeshType::VertexSet;
+  using VertexIndirectionSet =
+    slam::CoreArrayIndirectionSet<typename VertexSet::PositionType, axom::IndexType>;
+
+private:
+  /**
+   *  \brief Helper struct for sorting input points in the Biased Randomized Incremental Order (BRIO)
+   *
+   *  BRIO helps improve worst-case performance on poorly ordered point sets.
+   *  It was introduced in the following paper:
+   *    N. Amenta, S. Choi, and G. Rote. "Incremental constructions con BRIO."
+   *    Proceedings of the 19th annual symposium on Computational geometry, 2003.
+   */
+  struct BrioComparator
+  {
+    axom::IndexType m_index;
+    int m_level;
+    MortonIndexType m_morton;
+
+    BrioComparator(IndexType index, int level, MortonIndexType morton)
+      : m_index(index)
+      , m_level(level)
+      , m_morton(morton)
+    { }
+
+    friend bool operator<(const BrioComparator& lhs, const BrioComparator& rhs)
+    {
+      return (lhs.m_level == rhs.m_level) ? lhs.m_morton < rhs.m_morton
+                                          : lhs.m_level < rhs.m_level;
+    }
+  };
+
+  /**
+   * \brief Generates a permutation of [0, pts.size()) following BRIO 
+   * 
+   * \sa BrioComparator
+   */
+  template <typename PointArray>
+  axom::Array<axom::IndexType> computeInsertionOrder(const PointArray& pts,
+                                                     const BoundingBoxType& bb)
+  {
+    // This function will compute a permutation of pts following BRIO.
+    // Each point gets a level from the computeLevel() lambda
+    // and a quantized Morton index from a rectangular lattice over the bounding box
+
+    const int npts = pts.size();
+    const int nlevels =
+      axom::utilities::ceil(axom::utilities::log2<double>(npts));
+
+    // Each point has a 50% chance of being at the max level; of the remaining points
+    // from the previous level, there's a 50% chance of being at the current level.
+    // Any remaining points are at level 0.
+    auto computeLevel = [nlevels]() {
+      for(int level = nlevels; level > 0; --level)
+      {
+        if(axom::utilities::random_real(0., 1.) <= 0.5)
+        {
+          return level;
+        }
+      }
+      return 0;
+    };
+
+    // We use a Morton index, quantized over the mesh bounding box to
+    // order the points on each level
+    using QuantizedCoordType = axom::uint32;
+    using MortonizerType =
+      spin::Mortonizer<QuantizedCoordType, MortonIndexType, DIM>;
+
+    constexpr int shift_bits = (DIM == 2) ? 31 : 21;
```

Comment:
should these "magic numbers" be replaced by named constants defined at the top of the file?

Reply:
The numbers are the max bits per dimension that fit into a 64 bit integer, i.e. `floor(64/DIM)`

It's only used in this one spot -- I'll add a comment to clarify.